### PR TITLE
Add handling for UpgradeCertificate

### LINKED ISSUE #2444: [Upgradability] - Implement logic to detect if the network has been upgraded
### What is this task and why do we need to work on it?

A node should only start using the latest version of the protocol after the network has been upgraded. We should implement logic that detects if this has happened, and switches to the latest protocol version if so.

### What work will need to be done to complete this task?

- [ ] Create a global mutable variable that is updated when an upgrade certificate is encountered.
- [ ] Incoming messages with higher version numbers should be discarded.

### Are there any other details to include?

_No response_

### What are the acceptance criteria to close this issue?

Subtasks should be completed. Designing tests is likely to be complicated and should be a separate task.

### Branch work will be merged to (if not the default branch)

_No response_

### LINKED ISSUE #2445: [Upgradability] - Implement mechanism to initiate a network upgrade
### What is this task and why do we need to work on it?

In our current vision for the upgrade process, the leader for a round will begin collecting signatures on an upgrade certificate a few rounds in advance (which all nodes will then vote on). 

### What work will need to be done to complete this task?

_No response_

### Are there any other details to include?

_No response_

### What are the acceptance criteria to close this issue?

There should be a mechanism for:

- [ ] Initiating the upgrade process as an upcoming leader for a round.
- [ ] Voting on an upgrade certificate

### Branch work will be merged to (if not the default branch)

_No response_

### LINKED ISSUE #2556: [Upgradability] - Handle upgrade certificate in quorum proposal
### What is this task and why do we need to work on it?

Leaders collect signatures for an upgrade in an `UpgradeCertificate`, which should be attached to and handled with any `QuorumProposal` that goes out.

### What work will need to be done to complete this task?

_No response_

### Are there any other details to include?

_No response_

### What are the acceptance criteria to close this issue?

The leader for a view should be able to attach an upgrade certificate to a quorum proposal to initiate an upgrade. Nodes voting on an upgrade should validate this certificate, if a quorum proposal has one attached.

### Branch work will be merged to (if not the default branch)

_No response_

### PR DESCRIPTION
**Pull request by ss-es**
Closes #2556
Closes #2444
Closes #2445 

### This PR: 
Adds `UpgradeCertificate` handling, implementing the remaining upgradability tasks.

- Leaders can collect signatures on an `UpgradeCertificate` (done in [PR#2532](https://github.com/EspressoSystems/HotShot/pull/2532)), which is published as an `UpgradeCertificateFormed` event to the consensus task.
- If an `UpgradeCertificate` has been formed by the time the leader proposes, it is attached to the quorum proposal.
- Nodes validate any attached `UpgradeCertificate` before voting.
- Any new decided leaf with an `UpgradeCertificate` attached updates the consensus state with that certificate.

### Key places to review: 
The `upgrade` and `consensus` tasks in `crates/task-impls`.

### CODE REVIEW

File: crates/task-impls/src/consensus.rs [hunk 1]
```diff
@@ -496,6 +507,28 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
...
+                //
+                // We should just make sure we don't *sign* an UpgradeCertificate for an upgrade
+                // that we do not support.
+                if let Some(ref upgrade_cert) = proposal.data.upgrade_certificate {
+                    if !upgrade_cert.is_valid_cert(self.quorum_membership.as_ref()) {
+                        error!("Invalid upgrade_cert in proposal for view {}", *view);
+                        return;
+                    }
+                }
+
```
Comment (bfish713) [inline crates/task-impls/src/consensus.rs:531]: I think this is the correct logic as an honest node.  An honest node should not try to subvert the system by using a proposal as as second round of voting

File: crates/task-impls/src/consensus.rs [hunk 2]
```diff
@@ -727,6 +760,9 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
                                         .last_synced_block_height
                                         .set(usize::try_from(leaf.get_height()).unwrap_or(0));
                                 }
+                                if let Some(ref upgrade_cert) = proposal.data.upgrade_certificate {
+                                    self.decided_upgrade_cert = Some(upgrade_cert.clone());
+                                }
```
Comment (bfish713) [inline crates/task-impls/src/consensus.rs:765]: Nit: I think we should at least log this event at info level with the details of the upgrade cert

File: crates/task-impls/src/consensus.rs [hunk 3]
```diff
@@ -1274,12 +1331,27 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
...
+            if let Some(cert) = self.upgrade_cert.clone() {
+                if cert.view_number == view {
+                    upgrade_cert = Some(cert);
+                } else {
+                    upgrade_cert = None;
+                }
+            } else {
+                upgrade_cert = None;
+            }
+
```
Comment (bfish713) [inline crates/task-impls/src/consensus.rs:1347]: nit can probably remove the clone with something like this:
```suggestion

            let upgrade_cert = self.upgrade_cert.is_some_and(|cert| { cert.view_number == view }) {
                self.upgrade_cert.take()
            } else {
               None
            };

```

Also will leave self.upgrade_cert == None after we use it

**Review by bfish713 — APPROVED** [review]
All makes sense besides the few minor comments

File: crates/task-impls/src/consensus.rs (hunk 2 above)
Comment (ss-es) [reply crates/task-impls/src/consensus.rs]: Oh, yes -- this is definitely worth logging. Added!

File: crates/task-impls/src/consensus.rs (hunk 3 above)
Comment (ss-es) [reply crates/task-impls/src/consensus.rs]: Changed!

File: crates/task-impls/src/consensus.rs [hunk 4]
```diff
@@ -1274,12 +1332,29 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
...
+            let upgrade_cert = if self
+                .upgrade_cert
+                .as_ref()
+                .is_some_and(|cert| cert.view_number == view)
+            {
+                // If the cert view number matches, set upgrade_cert to self.upgrade_cert
+                // and set self.upgrade_cert to None.
+                //
+                // Note: the certificate is discarded, regardless of whether the vote on the proposal succeeds or not.
```
Comment (bfish713) [inline crates/task-impls/src/consensus.rs:1344]: I assumed this was the desired behavior, otherwise we can simply switch take() to clone()

**Review by bfish713 — APPROVED** [review]

File: crates/task-impls/src/consensus.rs (hunk 4 above)
Comment (ss-es) [reply crates/task-impls/src/consensus.rs:1344]: Yes, I think so! Just wanted to make it explicit.

**Pull request merged**